src/raw/mod.rs

```diff
@@ -862,52 +894,61 @@ impl<'f, A: Automaton> StreamWithState<'f, A> {
                     return;
                 }
             }
         }
         if !self.stack.is_empty() {
             let last = self.stack.len() - 1;
+            let state = &self.stack[last];
+            let transition = if !state.done { 
+                self.previous_transition(&state.node, state.trans)
+            } else {
+                self.last_transition(&state.node)
+            };
             if inclusive {
-                self.stack[last].trans -= 1;
+                self.stack[last].trans = transition.unwrap_or_default();
+                self.stack[last].done = transition.is_none();
                 self.inp.pop();
             } else {
-                let node = self.stack[last].node;
-                let trans = self.stack[last].trans;
+                let next_node = self.fst.node(state.node.transition(transition.unwrap_or_default()).addr, self.data);
+                let starting_transition = self.starting_transition(&next_node);
                 self.stack.push(StreamState {
-                    node: self.fst.node(node.transition(trans - 1).addr, self.data),
-                    trans: 0,
+                    node: next_node,
+                    trans: starting_transition.unwrap_or_default(),
                     out,
                     aut_state,
+                    done: starting_transition.is_none(),
                 });
             }
         }
     }
 
     #[inline]
     fn next<F, T>(&mut self, transform: F) -> Option<(&[u8], Output, T)> where F: Fn(&A::State) -> T {
-        if let Some(out) = self.empty_output.take() {
-            if self.end_at.exceeded_by(&[]) {
-                self.stack.clear();
-                return None;
-            }
-            let start = self.aut.start();
-            if self.aut.is_match(&start) {
-                return Some((&[], out, transform(&start)));
+        if !self.has_seeked {
+            self.seek();
+            self.has_seeked = true;
+        }
+        if !self.reversed {
+            // Inorder empty output (will be first).
+            if let Some(out) = self.empty_output.take() {
+                return self.resolve_empty_output(out, transform);
             }
         }
         while let Some(state) = self.stack.pop() {
-            if state.trans >= state.node.len() || !self.aut.can_match(&state.aut_state) {
+            if state.done || !self.aut.can_match(&state.aut_state) {
                 if state.node.addr() != self.fst.root_addr {
+                    self.inp_return = self.inp.clone();
                     self.inp.pop().unwrap();
+                    // Reversed return next logic.
+                    // If the stack is empty the value should not be returned. 
+                    if self.reversed && self.stack.len() > 0 && state.node.is_final() && 
+                      !self.out_of_bounds(&self.inp_return) && self.aut.can_match(&state.aut_state) {
```

Comment:
I am surprised this is `can_match` here, as opposed to `is_match`... Is this correct?

Reply:
Changed it and added unit tests.